### PR TITLE
Add missing interfaces of `torch.optim.swa_utils`

### PR DESCRIPTION
**Pull request by david20571015**
Add type hints for the function/class interfaces that appear in torch/optim/swa_utils.py but are missing in torch/optim/swa_utils.pyi.

- get_ema_multi_avg_fn
- get_swa_multi_avg_fn
- get_ema_avg_fn
- get_swa_avg_fn
- AveragedModel.__init__(multi_avg_fn)
- SWALR.get_lr

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/optim/swa_utils.pyi [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
+def get_ema_avg_fn(decay: float) -> Callable[[Tensor, Tensor, Any], Tensor]: ...
+
+def get_swa_avg_fn() -> Callable[[Tensor, Tensor, int], Tensor]: ...
+
 class AveragedModel(Module):
     def __init__(
         self,
         model: Module,
         device: Union[int, device] = ...,
         avg_fn: Callable[[Tensor, Tensor, int], Tensor] = ...,
```
Comment (janeyx99) [inline torch/optim/swa_utils.pyi:23]: Looking at the code (looking for `self.avg_fn(` and `self.multi_avg_fn(`) shows that the third arg should really be a scalar Tensor and not an int.

**Review by janeyx99 — COMMENTED** [review]
Thank you for creating this PR! We are attempting to move away from the py/pyi split and would prefer having the type hints exist on the original swa_utils.py themselves. 

What do you think about combining these types onto the .py file?

**Comment by david20571015** [comment]
Hello, thanks for your review. I'm working on combining both files, and I was wondering if there is a way to apply formatting?

**Comment by janeyx99** [comment]
@david20571015 PyTorch uses lintrunner: https://pypi.org/project/lintrunner/. You can install it and locally run `lintrunner init` followed by `lintrunner -a <filename>` _before_ committing to autoformat the file. Let me know if this didn't address the question!

**Comment by david20571015** [comment]
Thanks @janeyx99, it works well.

**Review by janeyx99 — COMMENTED** [review]
Tentatively looks good, gonna wait on CI

File: torch/optim/swa_utils.py [hunk 2]
```diff
@@ -162,30 +167,33 @@ class AveragedModel(Module):
...
         https://paperswithcode.com/method/polyak-averaging
     """
-    def __init__(self, model, device=None, avg_fn=None, multi_avg_fn=None, use_buffers=False):
+
+    def __init__(
+        self,
+        model: Module,
+        device: Optional[Union[int, device]] = None,
+        avg_fn: Optional[Callable[[Tensor, Tensor, Tensor], Tensor]] = None,
+        multi_avg_fn: Optional[Callable[[Tensor, Tensor, Tensor], Tensor]] = None,
```
Comment (janeyx99) [inline torch/optim/swa_utils.py:176]: These look wrong

Comment (david20571015) [reply torch/optim/swa_utils.py]: Fixed.
I also see some lint error from the CI, should I fix them?

**Comment by david20571015** [comment]
@janeyx99, I've fixed the typing of `multi_avg_fn`, and I've also noticed some error in the CI, such as `List item 0 has incompatible type "list[Tensor]"; expected "list[Tensor | None]"`.
Is this expected, or should I [fix](https://stackoverflow.com/questions/76765523/python-listt-not-assignable-to-listt-none) it?

**Comment by david20571015** [comment]
Hi @janeyx99, apologies for tagging you. Could you please check if my patch for your review is acceptable? Thank you.

File: torch/optim/swa_utils.py [hunk 3]
```diff
@@ -362,16 +374,17 @@ def _get_initial_lr(lr, swa_lr, alpha):
 
     def get_lr(self):
         if not self._get_lr_called_within_step:
-            warnings.warn("To get the last learning rate computed by the scheduler, "
-                          "please use `get_last_lr()`.", UserWarning)
+            warnings.warn(
```
Comment (janeyx99) [inline torch/optim/swa_utils.py:377]: how come there are so many spacing updates in this PR? would it be possible to keep the spacing the way it was?

**Review by janeyx99 — COMMENTED** [review]
Thanks for fixing this up--this looks much closer! Left some last thoughts.

Regarding the lint, I don't think changing List[Tensor] to List[Optional[Tensor]] is what we want here

File: torch/optim/swa_utils.py [hunk 4]
```diff
@@ -36,7 +41,7 @@ def ema_update(ema_param_list, current_param_list, _):
 
 def get_swa_multi_avg_fn():
     @torch.no_grad()
-    def swa_update(averaged_param_list, current_param_list, num_averaged):
+    def swa_update(averaged_param_list: PARAM_LIST, current_param_list: PARAM_LIST, num_averaged: Tensor):
```
[the remark this answers is not in the document]
Comment (david20571015) [reply torch/optim/swa_utils.py]: fixed

File: torch/optim/swa_utils.py [hunk 5]
```diff
@@ -49,15 +54,15 @@ def swa_update(averaged_param_list, current_param_list, num_averaged):
...
+    def ema_update(ema_param: Tensor, current_param: Tensor, num_averaged):
         return decay * ema_param + (1 - decay) * current_param
 
     return ema_update
 
 
 def get_swa_avg_fn():
     @torch.no_grad()
-    def swa_update(averaged_param, current_param, num_averaged):
+    def swa_update(averaged_param: Tensor, current_param: Tensor, num_averaged: Tensor):
```
[the remark this answers is not in the document]
Comment (david20571015) [reply torch/optim/swa_utils.py]: fixed

File: torch/optim/swa_utils.py [hunk 6]
```diff
@@ -162,30 +167,35 @@ class AveragedModel(Module):
...
-    def __init__(self, model, device=None, avg_fn=None, multi_avg_fn=None, use_buffers=False):
+
+    def __init__(
+        self,
+        model: Module,
+        device: Optional[Union[int, torch.device]] = None,
+        avg_fn: Optional[Callable[[Tensor, Tensor, Tensor], Tensor]] = None,
+        multi_avg_fn: Optional[
+            Callable[[List[Optional[Tensor]], List[Optional[Tensor]], Tensor],
+                     Tensor]] = None,
```
[the remark this answers is not in the document]
Comment (david20571015) [reply torch/optim/swa_utils.py]: Fixed, and also changed the type of the first two Tensor to `Union[Tuple[Tensor, ...], List[Tensor]]`.

File: torch/optim/swa_utils.py [hunk 7]
```diff
@@ -162,30 +167,35 @@ class AveragedModel(Module):
     .. _Polyak averaging:
         https://paperswithcode.com/method/polyak-averaging
     """
-    def __init__(self, model, device=None, avg_fn=None, multi_avg_fn=None, use_buffers=False):
+
+    def __init__(
+        self,
+        model: Module,
+        device: Optional[Union[int, torch.device]] = None,
+        avg_fn: Optional[Callable[[Tensor, Tensor, Tensor], Tensor]] = None,
```
[the remark this answers is not in the document]
Comment (david20571015) [reply torch/optim/swa_utils.py]: fixed

File: torch/optim/swa_utils.py (hunk 3 above)
Comment (david20571015) [reply torch/optim/swa_utils.py:377]: Fixed, this might cause by my wrong formatter config.

**Review by janeyx99 — COMMENTED** [review]
Looks better! Running CI to see what lint says.

Could you undo the formatting changes? They're very distracting from the actual changes!

**Review by janeyx99 — APPROVED** [review]
nice!

**Comment by janeyx99** [comment]
I'd merge, but it looks like running lintrunner is failing on the previous commit https://hud.pytorch.org/pr/117036

Looks like we may need to update some more type hints or add type ignores

**Comment by david20571015** [comment]
For each lint error, I have few questions. Could you give me some idea?

1. Should I construct a Tensor if `num_averaged` is an int, or will `torch._foreach_addcdiv_` do it?
```shell
Error (MYPY) [list-item]
  List item 0 has incompatible type "Tensor | int"; expected "Tensor"

       47  |            torch._foreach_lerp_(averaged_param_list, current_param_list, 1 / (num_averaged + 1))
       48  |        else:
       49  |            diffs = torch._foreach_sub(current_param_list, averaged_param_list)
  >>>  50  |            torch._foreach_addcdiv_(averaged_param_list, diffs, [num_averaged + 1] * len(averaged_param_list))
       51  |
       52  |    return swa_update
       53  |
```

2. Should I [cast](https://docs.python.org/3/library/typing.html#typing.cast) them, or use `#type: ignore`?
```shell
Error (MYPY) [list-item]
  List item 0 has incompatible type "list[Tensor]"; expected "list[Tensor
  | None]"

      207  |
      208  |        if self.n_averaged > 0:
      209  |            if self.multi_avg_fn is not None or self.avg_fn is None:
  >>> 210  |                grouped_tensors = _group_tensors_by_device_and_dtype([self_param_detached, model_param_detached])
      211  |                for (device, _), ([self_params, model_params], _) in grouped_tensors.items():
      212  |                    if self.multi_avg_fn:
      213  |                        self.multi_avg_fn(self_params, model_params, self.n_averaged.to(device))

Error (MYPY) [list-item]
  List item 1 has incompatible type "list[Tensor]"; expected "list[Tensor
  | None]"

      207  |
      208  |        if self.n_averaged > 0:
      209  |            if self.multi_avg_fn is not None or self.avg_fn is None:
  >>> 210  |                grouped_tensors = _group_tensors_by_device_and_dtype([self_param_detached, model_param_detached])
      211  |                for (device, _), ([self_params, model_params], _) in grouped_tensors.items():
      212  |                    if self.multi_avg_fn:
      213  |                        self.multi_avg_fn(self_params, model_params, self.n_averaged.to(device))

Error (MYPY) [arg-type]
  Argument 1 has incompatible type "list[Tensor | None]"; expected
  "tuple[Tensor, ...] | list[Tensor]"

      210  |                grouped_tensors = _group_tensors_by_device_and_dtype([self_param_detached, model_param_detached])
      211  |                for (device, _), ([self_params, model_params], _) in grouped_tensors.items():
      212  |                    if self.multi_avg_fn:
  >>> 213  |                        self.multi_avg_fn(self_params, model_params, self.n_averaged.to(device))
      214  |                    elif device.type in _get_foreach_kernels_supported_devices():
      215  |                        multi_avg_fn = get_swa_multi_avg_fn()
      216  |                        multi_avg_fn(self_params, model_params, self.n_averaged.to(device))

Error (MYPY) [arg-type]
  Argument 2 has incompatible type "list[Tensor | None]"; expected
  "tuple[Tensor, ...] | list[Tensor]"

      210  |                grouped_tensors = _group_tensors_by_device_and_dtype([self_param_detached, model_param_detached])
      211  |                for (device, _), ([self_params, model_params], _) in grouped_tensors.items():
      212  |                    if self.multi_avg_fn:
  >>> 213  |                        self.multi_avg_fn(self_params, model_params, self.n_averaged.to(device))
      214  |                    elif device.type in _get_foreach_kernels_supported_devices():
      215  |                        multi_avg_fn = get_swa_multi_avg_fn()
      216  |                        multi_avg_fn(self_params, model_params, self.n_averaged.to(device))

Error (MYPY) [assignment]
  Incompatible types in assignment (expression has type "Tensor | None",
  variable has type "Tensor")

      217  |                    else:
      218  |                        avg_fn = get_swa_avg_fn()
      219  |                        n_averaged = self.n_averaged.to(device)
  >>> 220  |                        for p_averaged, p_model in zip(self_params, model_params):
      221  |                            p_averaged.copy_(avg_fn(p_averaged, p_model, n_averaged))
      222  |            else:
      223  |                for p_averaged, p_model in zip(self_param_detached, model_param_detached):
```

3. Should I change the type of `momentum` from `float` to `Optional[float]`?
```shell
Error (MYPY) [assignment]
  Incompatible types in assignment (expression has type "None", variable has
  type "float")

      271  |    was_training = model.training
      272  |    model.train()
      273  |    for module in momenta.keys():
  >>> 274  |        module.momentum = None
      275  |
      276  |    for input in loader:
      277  |        if isinstance(input, (list, tuple)):
```

4. It seems that other derived classes of `LRScheduler` have the same usage. But imypy raise an error here.
```shell
Error (MYPY) [attr-defined]
  "SWALR" has no attribute "_get_lr_called_within_step"

      373  |        return (lr - alpha * swa_lr) / (1 - alpha)
      374  |
      375  |    def get_lr(self):
  >>> 376  |        if not self._get_lr_called_within_step:
      377  |            warnings.warn("To get the last learning rate computed by the scheduler, "
      378  |                          "please use `get_last_lr()`.", UserWarning)
      379  |        step = self._step_count - 1
```

5. `_step_count` was set [here](https://github.com/pytorch/pytorch/blob/67a4d6d6cb421a6eeb42ec6ca217500a11abe201/torch/optim/lr_scheduler.py#L90), but mypy can't recognize it.
```shell
Error (MYPY) [attr-defined]
  "SWALR" has no attribute "_step_count"
      376  |        if not self._get_lr_called_within_step:
      377  |            warnings.warn("To get the last learning rate computed by the scheduler, "
      378  |                          "please use `get_last_lr()`.", UserWarning)
  >>> 379  |        step = self._step_count - 1
      380  |        if self.anneal_epochs == 0:
      381  |            step = max(1, step)
      382  |        prev_t = max(0, min(1, (step - 1) / max(1, self.anneal_epochs)))
```

**Comment by janeyx99** [comment]
Thank you for your due diligence here!
1. Oh...I see. It would be more performant if you branched to use regular _foreach_add with an alpha of the reciprocal if num_average is not a Tensor (though you'd need to convert to double). 
2. casting seems fine!
3. yes
4. ignore with a comment explaining how it's dynamically set
5. same as 4

File: torch/optim/swa_utils.py [hunk 8]
```diff
@@ -47,7 +47,10 @@ def swa_update(averaged_param_list: PARAM_LIST, current_param_list: PARAM_LIST,
             torch._foreach_lerp_(averaged_param_list, current_param_list, 1 / (num_averaged + 1))
         else:
             diffs = torch._foreach_sub(current_param_list, averaged_param_list)
-            torch._foreach_addcdiv_(averaged_param_list, diffs, [num_averaged + 1] * len(averaged_param_list))
+            if isinstance(num_averaged, Tensor):
+                torch._foreach_addcdiv_(averaged_param_list, diffs, [num_averaged + 1] * len(averaged_param_list))
+            else:
+                torch._foreach_add_(averaged_param_list, diffs, alpha=1 / (num_averaged + 1))
```
Comment (janeyx99) [inline torch/optim/swa_utils.py:53]:
```suggestion
                torch._foreach_add_(averaged_param_list, diffs, alpha=1.0 / (num_averaged + 1))
```
if num_averaged is an int?

Comment (david20571015) [reply torch/optim/swa_utils.py]: Thanks for fixing this.

**Comment by david20571015** [comment]
The failed test does not seem to have been caused by me. Should I handle it?

**Comment by david20571015** [comment]
Hello @janeyx99, would you mind approving the lint CI for me, please?

**Comment by janeyx99** [comment]
If lint passes here I am going to just try to force merge this...this process is taking too long

**Pull request closed without merging**